crates/sui-core/src/storage.rs

```diff
@@ -29,6 +30,9 @@ pub struct RocksDbStore {
     authority_store: Arc<AuthorityStore>,
     committee_store: Arc<CommitteeStore>,
     checkpoint_store: Arc<CheckpointStore>,
+    // in memory checkpoint watermark sequence numbers
+    highest_verified_checkpoint: Arc<Mutex<Option<u64>>>,
+    highest_synced_checkpoint: Arc<Mutex<Option<u64>>>,
```

Comment:
I didn't find where we use these two values?

Reply:
Some code got deleted unintentionally, it's usage should be visible now

Reply:
I see them now. We do have a pre-write DB check in `CheckpointStore::update_highest_verified_checkpoint` and `update_highest_executed_checkpoint` (but not `update_highest_synced_checkpoint` though interestingly). I think it's better to keep the in-mem check  in `CheckpointStore` or perhaps we don't need the in-mem checks? Though we do want to add the checks in `update_highest_synced_checkpoint`

Reply:
Fair point Lu. But that could still allow for a racing read-modify-write to the watermark because it's not guarded with any lock (or other synchronization primitive). The reason I think it is important to have this here is this scenario we could still have where both state-sync-from-archive and state-sync-from-peers might end up overlapping in the checkpoint range they are downloading and updating watermarks (from concurrent running tokio tasks):

```
1. State sync peers announce the lowest checkpoint they have is 10 and highest checkpoint they have is 100. The lowest_peer_checkpoint_watermark is 10
2. Archive has checkpoint data from 0 to 50
3. We start catching up checkpoints from archive from sequence number 0 (and state sync from peers is just waiting) until checkpoint 10 (checkpoint 9) in tokio task 1
4. A new state sync peer joins or an existing peer backfills its own state such that lowest_peer_checkpoint_watermark moves back to 5
5. In such a scenario, state sync from peers will also start trying to download checkpoints in parallel (tokio task 2) and both tasks could race trying to perform read-modify-write on the checkpoint watermarks 
```

I think it would be quite rare but it is better to have this protection than not imo.